pkg/virt-api/rest/subresource.go

```diff
@@ -288,6 +281,28 @@ func (app *SubresourceAPIApp) StartVMRequestHandler(request *restful.Request, re
 		return
 	}
 
+	vmi, err := app.VirtCli.VirtualMachineInstance(namespace).Get(name, &k8smetav1.GetOptions{})
+
+	var (
+		vmiStatusPhase = v1.Unknown
+		vmiIsFinal     = false
+	)
+	if vmi != nil {
+		vmiStatusPhase = vmi.Status.Phase
+		vmiIsFinal = vmi.IsFinal()
+	}
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			response.WriteError(http.StatusInternalServerError, err)
+			return
+		}
+	} else {
+		if !vmiIsFinal {
```

Comment:
I don't think this will work, because you are setting `vmiFinal` to false yourself in case the VMI does not exist... but in that case we don't want to get in here. I'm not sure if using your own vars and setting them to some "random" default values is such a good idea. At least to me a nil check on `vmi` isn't that awkward ;)

Reply:
Yes, exactly, we don't want to go in there :)

My first idea on this was to eliminate the several nil checks being scattered all over the code to check whether `vmi != nil && vmi.State.Phase == ...` and to use a poor mans Null Object instead. Everything else was a series of refactorings which simplified the code and what you see here is what is left of that.

Actually the refactorings led me to see that

    if vmi != nil && vmi.State.Phase == v1.Successful || vmi != nil && vmi.State.Phase == v1.Failed

would be same as

    if vmi != nil && vmi.IsFinal()

would be same for (in pseudo Python)

    vmiIsFinal = if vmi is nil then false else vmi.IsFinal()
    ...
    if !vmiIsFinal

And then that was all that was left. Unfortunately this transformation is not visible here :(

Reply:
On second thought... I'm not sure if that case is covered, so I will investigate...